overlord/snapstate/storehelpers.go

```diff
@@ -63,19 +103,154 @@ func updateInfo(st *state.State, snapst *SnapState, channel string, ignoreValida
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if curInfo.SnapID == "" { // covers also trymode
+		return nil, nil, store.ErrLocalSnap
+	}
+
+	return curInfo, user, nil
+}
+
+func updateToRevisionInfo(st *state.State, snapst *SnapState, channel string, revision snap.Revision, userID int) (*snap.Info, error) {
+	curInfo, user, err := preUpdateInfo(st, snapst, userID)
 	if err != nil {
 		return nil, err
 	}
+
 	theStore := Store(st)
 	st.Unlock() // calls to the store should be done without holding the state lock
 	spec := store.SnapSpec{
-		Name:     name,
+		Name:     curInfo.Name(),
 		Channel:  channel,
 		Revision: revision,
 	}
 	snap, err := theStore.SnapInfo(spec, user)
 	st.Lock()
 	return snap, err
 }
+
+func refreshCandidates(st *state.State, names []string, user *auth.UserState, flags *store.RefreshOptions) ([]*snap.Info, map[string]*SnapState, map[string]bool, error) {
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	sort.Strings(names)
+
+	stateByID := make(map[string]*SnapState, len(snapStates))
+	candidatesInfo := make([]*store.RefreshCandidate, 0, len(snapStates))
+	ignoreValidation := make(map[string]bool)
+	userIDs := make(map[int]bool)
+	for _, snapst := range snapStates {
+		if len(names) == 0 && (snapst.TryMode || snapst.DevMode) {
+			// no auto-refresh for trymode nor devmode
+			continue
+		}
+
+		// FIXME: snaps that are not active are skipped for now
+		//        until we know what we want to do
+		if !snapst.Active {
+			continue
+		}
+
+		snapInfo, err := snapst.CurrentInfo()
+		if err != nil {
+			// log something maybe?
+			continue
+		}
+
+		if snapInfo.SnapID == "" {
+			// no refresh for sideloaded
+			continue
+		}
+
+		if len(names) > 0 && !strutil.SortedListContains(names, snapInfo.Name()) {
+			continue
+		}
+
+		stateByID[snapInfo.SnapID] = snapst
+
+		// get confinement preference from the snapstate
+		candidateInfo := &store.RefreshCandidate{
+			// the desired channel (not info.Channel!)
+			Channel:          snapst.Channel,
+			SnapID:           snapInfo.SnapID,
+			Revision:         snapInfo.Revision,
+			Epoch:            snapInfo.Epoch,
+			IgnoreValidation: snapst.IgnoreValidation,
+		}
+
+		if len(names) == 0 {
+			candidateInfo.Block = snapst.Block()
+		}
+
+		candidatesInfo = append(candidatesInfo, candidateInfo)
+		if snapst.UserID != 0 {
+			userIDs[snapst.UserID] = true
+		}
+		if snapst.IgnoreValidation {
+			ignoreValidation[snapInfo.SnapID] = true
+		}
+	}
+
+	theStore := Store(st)
+
+	// TODO: we query for all snaps for each user so that the
+	// store can take into account validation constraints, we can
+	// do better with coming APIs
+	updatesInfo := make(map[string]*snap.Info, len(candidatesInfo))
+	fallbackUsed := false
+	fallbackID := idForUser(user)
+	if len(userIDs) == 0 {
+		// none of the snaps had an installed user set, just
+		// use the fallbackID
+		userIDs[fallbackID] = true
+	}
+	for userID := range userIDs {
+		u := user
```

Comment:
Could we use a helper like userFromUserID here to make this and the following lines a bit more compact? It might be slightly easier to follow with the user getting extracted.

Reply:
I introduced userFromUserIDOrFallback